pkg/sync/parallel_syncer.go

```diff
@@ -0,0 +1,1706 @@
+type StateInterface interface {
+	PageToken(ctx context.Context) string
+	NextPage(ctx context.Context, pageToken string) error
+	SetNeedsExpansion()
+	SetHasExternalResourcesGrants()
+	ShouldFetchRelatedResources() bool
+}
+
+// ActionDecision represents the decision made by a helper method
+// This allows the caller to decide how to handle the result
+type ActionDecision struct {
+	ShouldContinue       bool   // Whether to continue processing (e.g., more pages)
+	NextPageToken        string // Page token for next page, if applicable
+	Action               string // What action to take: "next_page", "finish", "continue"
+	NeedsExpansion       bool   // Whether grant expansion is needed
+	HasExternalResources bool   // Whether external resources were found
+	ShouldFetchRelated   bool   // Whether related resources should be fetched
+}
+
+// LocalStateContext provides local state management for parallel syncer
+// This implements StateInterface without sharing the global state machine
+type LocalStateContext struct {
+	resourceID           *v2.ResourceId
+	pageToken            string
+	needsExpansion       bool
+	hasExternalResources bool
+	shouldFetchRelated   bool
+}
+
+// NewLocalStateContext creates a new local state context for a resource
+func NewLocalStateContext(resourceID *v2.ResourceId) *LocalStateContext {
+	return &LocalStateContext{
+		resourceID:           resourceID,
+		pageToken:            "",
+		needsExpansion:       false,
+		hasExternalResources: false,
+		shouldFetchRelated:   true, // Default to true for backward compatibility
+	}
+}
+
+// PageToken returns the current page token for this resource
+func (lsc *LocalStateContext) PageToken(ctx context.Context) string {
+	return lsc.pageToken
+}
+
+// NextPage updates the page token for the next page
+func (lsc *LocalStateContext) NextPage(ctx context.Context, pageToken string) error {
+	lsc.pageToken = pageToken
+	return nil
+}
+
+// SetNeedsExpansion marks that grant expansion is needed
+func (lsc *LocalStateContext) SetNeedsExpansion() {
+	lsc.needsExpansion = true
+}
+
+// SetHasExternalResourcesGrants marks that external resources were found
+func (lsc *LocalStateContext) SetHasExternalResourcesGrants() {
+	lsc.hasExternalResources = true
+}
+
+// ShouldFetchRelatedResources returns whether related resources should be fetched
+func (lsc *LocalStateContext) ShouldFetchRelatedResources() bool {
+	return lsc.shouldFetchRelated
+}
+
+// ParallelSyncConfig holds configuration for parallel sync operations
+type ParallelSyncConfig struct {
+	// Number of worker goroutines to use for parallel processing
+	WorkerCount int
+	// Default bucket for resource types not explicitly configured
+	// If a resource type doesn't specify a sync_bucket, this default will be used
+	DefaultBucket string
+}
+
+// DefaultParallelSyncConfig returns a default configuration
+func DefaultParallelSyncConfig() *ParallelSyncConfig {
+	return &ParallelSyncConfig{
+		WorkerCount:   1,
```

Comment:
Shouldn't we default to a number higher than 1?

Reply:
IMO whatever codepath gets us to partial sync should require a worker count as an arg that must be >= 2